syne_tune/callbacks/hyperband_remove_checkpoints_callback.py

```diff
@@ -0,0 +1,518 @@
+    :class:`HyperbandRemoveCheckpointsBaselineCallback`.
+    """
+
+    def __init__(
+        self,
+        max_num_checkpoints: int,
+        max_wallclock_time: int,
+        metric: str,
+        resource_attr: str,
+        mode: str,
+    ):
+        self.max_num_checkpoints = max_num_checkpoints
+        self._max_wallclock_time = max_wallclock_time
+        self._metric = metric
+        self._resource_attr = resource_attr
+        assert mode in ["min", "max"]
+        self._metric_sign = -1 if mode == "max" else 1
+        self._trials_with_checkpoints_removed = None
+        self._trial_status = None
+        self._trials_resumed_without_checkpoint = None
+        self._num_checkpoints_removed = None
+        self._num_trials_resumed = None
+        self._terminator = None
+        self._trial_backend = None
+
+    def _check_and_initialize(self, tuner):
+        scheduler = tuner.scheduler
+        self._terminator = scheduler.terminator
+        self._trial_backend = tuner.trial_backend
+        assert not isinstance(
+            self._trial_backend, SimulatorBackend
+        ), "Checkpoint removal not supported for SimulatorBackend"
+
+    def on_tuning_start(self, tuner):
+        self._check_and_initialize(tuner)
+        # For any paused trial with checkpoint removed, this dict maps ``trial_id`` to
+        # ``level`` (where this trial is paused)
+        self._trials_with_checkpoints_removed = dict()
+        # Maps ``trial_id`` to ``TrialStatus``. Used to count the number of
+        # trials with checkpoints
+        self._trial_status = dict()
+        # Collects entries ``(trial_id, level)`` from ``_trials_with_checkpoints_removed``
+        # which were nevertheless resumed
+        self._trials_resumed_without_checkpoint = []
+        self._num_checkpoints_removed = 0
+        self._num_trials_resumed = 0
+
+    @property
+    def num_checkpoints_removed(self) -> int:
+        return self._num_checkpoints_removed
+
+    def _filter_paused_trials(
+        self, paused_trials: PausedTrialsResult
+    ) -> PausedTrialsResult:
+        return [
+            entry
+            for entry in paused_trials
+            if entry[0] not in self._trials_with_checkpoints_removed
+        ]
+
+    def _count_trials_with_checkpoints(self) -> int:
+        has_checkpoint = {TrialStatus.RUNNING, TrialStatus.PAUSED_WITH_CHECKPOINT}
+        return sum(status in has_checkpoint for status in self._trial_status.values())
+
+    def _remove_checkpoint_of(self, trial_id: str, level: int):
+        self._trial_backend.delete_checkpoint(int(trial_id))
+        self._trial_status[trial_id] = TrialStatus.PAUSED_NO_CHECKPOINT
+        self._trials_with_checkpoints_removed[trial_id] = level
+        self._num_checkpoints_removed += 1
+
+    def _trials_to_be_removed(
+        self,
+        paused_trials_with_checkpoints: PausedTrialsResult,
+        num_to_remove: int,
+    ) -> List[Tuple[str, int, int, int, float]]:
```

Comment:
We should introduce the Trial data structure that can hold this kind of information (like below). It should be a dataclass with default values so it can be used to pass information around even if not all entries are known as in `trials_to_score` in L335

```
class TrialInfo
    trial_id: str= None
    level: int= None
    rank: int= None
    rung_len: int= None
    score: float= None
```

Reply:
OK